wikipron/extract/jpn.py

```diff
@@ -0,0 +1,229 @@
+                ]
+            ]
+
+]
+"""
+
+# Check for more than one etymology for the same reason as in lat.py
+# If there is more than one etymology, the headings get shifted down a level.
+_TOC_ETYMOLOGY_XPATH_SELECTOR = """
+//a[@href = "#Japanese"]
+  /following-sibling::ul
+    /li
+      /a
+        [starts-with(@href, "#Etymology")]
+"""
+
+
+# Some pages may not have etymology section.
+def _check_etymologies(request: requests.Response):
+    count = 0
+    for a_element in request.html.xpath(_TOC_ETYMOLOGY_XPATH_SELECTOR):
+        count += 1
+    return "h4" if count > 1 else "h3"
+
+
+# Check if hiragana entry on page
+def _check_hiragana(request: requests.Response, hiragana_path: str):
+    try:
+        request.html.xpath(hiragana_path)[0]
+        return True
+    except IndexError:
+        return False
+
+
+# The extraction functions below speak to some of the complexity
+# involved with Japanese. In other languages, on each entry page,
+# we are always mapping prons to a single word. In Japanese an entry page
+# may contain multiple prons that need to be mapped to multiple words.
+# This extraction function avoids some of this complexity by always only
+# taking one word for each word-pron pairing we find on an entry page,
+# but it does this for as many pairings as we can find on the entry page.
+def _yield_jpn_word(
+    request: requests.Response, pair_check_path: str, word_path: str
+) -> "Iterator[Word]":
+    # Iterate through the pair groups, grabbing the
+    # first hiragana word in each.
+    for pair_group in request.html.xpath(pair_check_path):
+        try:
+            word_element = pair_group.xpath(word_path)[0]
+        except IndexError:
+            return
+        word = word_element.text.rstrip(",(")
+        yield word
+
+
+# Japanese "Pronunciation" headers are most often sibling to a <ul> containing
+# an IPA transcription. There may also be a <ul> sibling to that <ul> which
+# contains a second IPA transcription (though second transcriptions are not
+# always contained within a separate <ul>). The only way I could work in
+# grabbing the transcriptions in both <ul>'s with my attempt to connect
+# words and prons "locally" - by moving stepwise from the word to the pron -
+# was to update _PRON_XPATH_SELECTOR with the second <ul> and
+# run a separate request.
+def _yield_jpn_upper_pron(
+    request: requests.Response, config: "Config", word_target: str
+) -> "Iterator[Pron]":
+    pron_path = _PRON_XPATH_SELECTOR.format(
+        word_to_work_from=word_target,
+        heading=_check_etymologies(request),
+        second_ul="[preceding-sibling::*[1][self::ul]]",
+    )
+    try:
+        request.html.xpath(pron_path)[0]
+    except IndexError:
+        return
+
+    for upper_pron_ele in request.html.xpath(pron_path):
+        prons = []
+        for pron in yield_pron(upper_pron_ele, IPA_XPATH_SELECTOR, config):
+            prons.append(pron)
+        yield prons
```

Comment:
Would the following work?

```python
prons = list(yield_pron(upper_pron_ele, IPA_XPATH_SELECTOR, config))
yield prons
```

Reply:
can't you do `yield from yield_pron(...)`?

Reply:
IIUC, the code was trying to yield a list of prons, rather than yield the prons one by one, and hence my suggested refactoring. That being said, now that I'm reading this again in the larger context, I think Lucas was trying to generate both "lower" and "upper" prons, chain them together, and then yield the prons one at a time. In this case, I think there's a more natural way to refactor the code. The first part is what Kyle has suggested: We should do `yield from yield_pron(...)` here, so that `_yield_jpn_upper_pron` yields the prons one by one. Other parts are in my updated comments below.

Reply:
The idea in both of these pron yielding functions is that I want to exhaustively run `yield_pron()` on each iteration of the upper for loops in each of the pron yielding functions. 

When the next `next(prons)` in `_combine_word_pron_pairs()` gets called, the upper for loop in both pron yielding functions gets ‘incremented’ (if that’s the right word?) and we target the subsequent word-pron pair group on the page and grab the prons in that group.

I am trying to potentially yield a list of prons in each function. Jackson's original `list()` suggestion is a simplified way of exhaustively running `yield_pron()`, but it still leaves this:
```
        try:
            prons += next(upper_prons)
        except StopIteration:
            pass
```
block in `_yield_jpn_lower_pron`. (If this comment and the comment I left on the `itertools.chain()` thread don't make any sense then let me know.)